raft/raft_test.go

```diff
@@ -902,61 +869,55 @@ func TestDisruptiveFollower2AA(t *testing.T) {
-	storage := NewMemoryStorage()
-	storage.ApplySnapshot(s)
-	sm := newTestRaft(1, nil, 10, 1, storage)
-	sm.Term = 1
-
-	sm.becomeCandidate()
-	sm.becomeLeader()
-	sm.Step(pb.Message{MsgType: pb.MessageType_MsgPropose, Entries: []*pb.Entry{{}}})
-	sm.readMessages() // clear message
-	// slow follower
-	sm.Prs[2].Match, sm.Prs[2].Next = 5, 6
-	// normal follower
-	sm.Prs[3].Match, sm.Prs[3].Next = sm.RaftLog.LastIndex(), sm.RaftLog.LastIndex()+1
-
-	sm.Step(pb.Message{MsgType: pb.MessageType_MsgBeat})
-	msgs := sm.readMessages()
-	if len(msgs) != 2 {
-		t.Fatalf("len(msgs) = %v, want 2", len(msgs))
-	}
-	wantCommitMap := map[uint64]uint64{
-		2: min(sm.RaftLog.committed, sm.Prs[2].Match),
-		3: min(sm.RaftLog.committed, sm.Prs[3].Match),
+func TestHeartbeatUpdateCommit2AB(t *testing.T) {
+	tests := []struct {
+		failCnt    int
+		successCnt int
+	}{
+		{1, 1},
+		{5, 3},
+		{10, 10},
 	}
-	for i, m := range msgs {
-		if m.MsgType != pb.MessageType_MsgHeartbeat {
-			t.Fatalf("#%d: type = %v, want = %v", i, m.MsgType, pb.MessageType_MsgHeartbeat)
+	for i, tt := range tests {
+		sm1 := newTestRaft(1, []uint64{1, 2, 3}, 10, 1, NewMemoryStorage())
+		sm2 := newTestRaft(1, []uint64{1, 2, 3}, 10, 1, NewMemoryStorage())
+		sm3 := newTestRaft(1, []uint64{1, 2, 3}, 10, 1, NewMemoryStorage())
+		nt := newNetwork(sm1, sm2, sm3)
+		nt.send(pb.Message{From: 1, To: 1, MsgType: pb.MessageType_MsgHup})
+		nt.isolate(1)
+		// propose log to old leader should fail
+		for i := 0; i < tt.failCnt; i++ {
+			nt.send(pb.Message{From: 1, To: 1, MsgType: pb.MessageType_MsgPropose, Entries: []*pb.Entry{{}}})
 		}
-		if m.Index != 0 {
-			t.Fatalf("#%d: prevIndex = %d, want %d", i, m.Index, 0)
+		if sm1.RaftLog.committed > 1 {
+			t.Fatalf("#%d: unexpected commit: %d", i, sm1.RaftLog.committed)
 		}
-		if m.LogTerm != 0 {
-			t.Fatalf("#%d: prevTerm = %d, want %d", i, m.LogTerm, 0)
+		// propose log to cluster should success
+		nt.send(pb.Message{From: 2, To: 2, MsgType: pb.MessageType_MsgHup})
+		for i := 0; i < tt.successCnt; i++ {
+			nt.send(pb.Message{From: 2, To: 2, MsgType: pb.MessageType_MsgPropose, Entries: []*pb.Entry{{}}})
 		}
-		if wantCommitMap[m.To] == 0 {
-			t.Fatalf("#%d: unexpected to %d", i, m.To)
-		} else {
-			if m.Commit != wantCommitMap[m.To] {
-				t.Fatalf("#%d: commit = %d, want %d", i, m.Commit, wantCommitMap[m.To])
+		wCommit := uint64(2 + tt.successCnt) // 2 elctions
+		if sm2.RaftLog.committed != wCommit {
+			t.Fatalf("#%d: expected sm2 commit: %d, got: %d", i, wCommit, sm2.RaftLog.committed)
+		}
+		if sm3.RaftLog.committed != wCommit {
+			t.Fatalf("#%d: expected sm3 commit: %d, got: %d", i, wCommit, sm3.RaftLog.committed)
+		}
+
+		nt.recover()
+		nt.send(pb.Message{From: 2, To: 2, MsgType: pb.MessageType_MsgBeat})
+		committed := sm1.RaftLog.nextEnts()
+		for _, log := range committed {
+			if log.Index == 1 {
+				continue
+			}
+			if log.Term != sm2.Term {
+				t.Fatalf("#%d: sm1 should commit entrie from sm2, got: %v", i, log)
 			}
```

Comment:
```suggestion
		nt.recover()
		nt.ignore(pb.MessageType_MsgAppend)
		nt.send(pb.Message{From: 2, To: 2, MsgType: pb.MessageType_MsgBeat})
```
We should testing how`sm1`'s commit index update by `Heartbeat` instead of `AppendEntries`.